src/either_or_both.rs

```diff
@@ -8,3 +10,49 @@ pub enum EitherOrBoth<A, B> {
     /// Only the right value of type `B` is present.
     Right(B),
 }
+
+impl<A, B> EitherOrBoth<A, B> {
+    /// If `Left`, or `Both`, return true, otherwise, return false.
+    pub fn has_left(&self) -> bool {
+        self.as_ref().left().is_some()
+    }
+
+    /// If `Right`, or `Both`, return true, otherwise, return false.
+    pub fn has_right(&self) -> bool {
+        self.as_ref().right().is_some()
+    }
+
+    /// If `Left`, or `Both`, return `Some` with that value, otherwise, return `None`.
+    pub fn left(self) -> Option<A> {
+        match self {
+            Left(left) | Both(left, _) => Some(left),
+            _ => None
+        }
+    }
+
+    /// If `Right`, or `Both`, return `Some` with that value, otherwise, return `None`.
+    pub fn right(self) -> Option<B> {
+        match self {
+            Right(right) | Both(_, right) => Some(right),
+            _ => None
+        }
+    }
+
+    /// Converts from `EitherOrBoth<A, B>` to `EitherOrBoth<&A, &B>`.
+    pub fn as_ref(&self) -> EitherOrBoth<&A, &B> {
+        match self {
```

Comment:
Style wise this is preferred as `*self` and dropping the `&` on each line below. Same with mut. But that's not important enough to block merging.